[APM] Average page load duration by country chart

ogupte opened the pull request:
Adds the map chart from #35571

<img width="1006" alt="Screen Shot 2019-08-29 at 2 08 24 PM" src="https://user-images.githubusercontent.com/1967266/63976601-89453580-ca66-11e9-8482-68f15c25129e.png">

Can test locally by generating interesting geo data for the opbeans-rum app:
Use data generation tool at https://github.com/ogupte/apm-ui-seed
`apm-ui-seed-geo --host='localhost:9200' --index='apm-8.0.0-transaction' --service-name='client' clean seed list`. Then view page-loads for a RUM service.

ogupte commented:
screenshot with tooltip:
<img width="628" alt="Screen Shot 2019-08-19 at 10 30 53 PM" src="https://user-images.githubusercontent.com/1967266/63328980-f11cb300-c2e5-11e9-9bf8-2d221fc00abc.png">

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/13276/)

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+  return `hsl(${color.hue},${color.saturation}%,${lightness}%)`;
+}
+
+const RegionMapChartToolTip: React.SFC<{
```
@sqren pointed out that there are some issues with the `React.(S)FC` type (implicit typing for children, default props and some other stuff too), and that it's often better to just type `Props`. Can't find the comment immediately though.

sqren replied:
I'm not strongly against SFC, but since it has a slightly more clunky syntax and comes with a few caveats I avoid it. But if you prefer it I'm fine with that too :)

At some point I wondered whether other agreed with me on this and found this little write-up: https://github.com/typescript-cheatsheets/react-typescript-cheatsheet#function-components (section: `What about `React.FC`/`React.FunctionComponent`?`)

Again: not deal-breaker at all. But that was my reasoning :)

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+function getProgressionColor(
+  scale: number,
+  // default to euiColorPrimary (#006BB4)
+  color = { hue: 204, saturation: 100, lightness: { min: 20, max: 90 } }
```
perhaps you can use `hexToHsv` from EUI?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+function getProgressionColor(
+  scale: number,
+  // default to euiColorPrimary (#006BB4)
+  color = { hue: 204, saturation: 100, lightness: { min: 20, max: 90 } }
```
perhaps you can use hexToHsv from EUI?

sqren replied:
`polished` is the lodash of styling and we've used it elsewhere to do similar things: https://polished.js.org/docs/

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+  data?: { key: string; value: number } & Partial<{ docCount: number }>;
+}> = ({ geojsonProperties, data }) => {
+  return (
+    <div style={{ textAlign: 'center' }}>
```
What's our policy re: inline styles?

sqren replied:
I haven't put too much thought into this but sometimes I also use inline styling instead of creating a separate "styled-components" component. For instance if I only need to add a single style, and the component is not re-used anywhere else. But perhaps it's better to go ahead and create the rule like we normally would? What do you think?

If we upgrade to the latest version of styled-components we'll get access to the `css` prop which feels like inline styling but is turned into a styled-component under the hood.
https://medium.com/styled-components/announcing-native-support-for-the-css-prop-in-styled-components-245ca5252feb

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+  return (
+    <div style={{ textAlign: 'center' }}>
+      <div style={{ fontSize: 'larger' }}>{geojsonProperties.name}</div>
+      <div>Avg. page load duration:</div>
```
translation?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+      <div style={{ fontWeight: 'bold', fontSize: 'larger' }}>
+        {data && data.value}ms
+      </div>
+      <div>({data && data.docCount} page loads)</div>
```
same here: translation?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+  const { data } = useAvgDurationByCountry();
+  const choroplethData = useMemo(
+    () =>
+      data.map(
```
Maybe this transformation should be in the endpoint as well? Considering that we're not using any UI state.

sqren replied:
+1

dgieselaar commented on x-pack/legacy/plugins/apm/public/hooks/useAvgDurationByCountry.ts:
```diff
@@ -0,0 +1,45 @@
+    urlParams: {
+      serviceName,
+      start,
+      end /* transactionName, transactionType*/
```
do these comments mean that these parameters will get implemented later, or will they be removed?

dgieselaar commented on x-pack/legacy/plugins/apm/public/services/rest/apm/transaction_groups.ts:
```diff
@@ -121,3 +122,24 @@ export async function loadTransactionBreakdown({
+  end: string;
+  uiFilters: UIFilters;
+}) {
+  return callApi<TransactionAvgDurationByCountryAPIResponse>({
```
You can use `callApmApi` here (after pulling in from `master`).

dgieselaar commented on x-pack/legacy/plugins/apm/server/routes/transaction_groups.ts:
```diff
@@ -146,4 +147,23 @@ export function initTransactionGroupsApi(core: InternalCoreSetup) {
     }
   });
+
+  server.route({
```
This would need to change as well after you pull in the route changes from `master`.

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,76 @@
+
+const RegionMapChartToolTip: React.SFC<{
+  geojsonProperties: { [name: string]: any };
+  data?: { key: string; value: number } & Partial<{ docCount: number }>;
```
out of interest, why use `Partial<{}>` here and not `{key:string, value: number, docCount?: number}`?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ColorProgressionBar.tsx:
```diff
@@ -0,0 +1,40 @@
+export const ColorProgressionBar: React.SFC<{
+  slices: number;
+  getColorStyle: (scale: number) => string;
+  showProgressiveHeight?: boolean;
```
are we using this right now? should we add it if we don't?

dgieselaar commented on x-pack/dev-tools/jest/setup/polyfills.js:
```diff
@@ -13,3 +13,6 @@ bluebird.Promise.setScheduler(function (fn) { global.setImmediate.call(global, f
 const MutationObserver = require('mutation-observer');
 Object.defineProperty(window, 'MutationObserver', { value: MutationObserver });
+
+const URL = { createObjectURL: () => '' };
```
Why do we need this? (Maybe good to leave a comment here anyway).

dgieselaar commented:
Tooltip looks strange in Chrome/OS X:

![image](https://user-images.githubusercontent.com/352732/63636678-af3b9600-c672-11e9-8693-1b3e933415c0.png)

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+  const hoveredFeaturePropertiesRef = useRef<
+    mapboxgl.MapboxGeoJSONFeature['properties']
+  >(null);
+  const isMapReady = useRef<boolean>(false);
```
Could you explain why you're using `useRef` instead of `useState`/`useCallback`? I've got a feeling this code would be simpler and more robust if we avoid useRef.

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+      if (currentLayer) {
+        map.removeLayer(choroplethLayerId.current);
+      }
+      const symbolLayer = (map.getStyle().layers || []).find(
```
Are we using this layer?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+      const symbolLayer = (map.getStyle().layers || []).find(
+        ({ type }) => type === 'symbol'
+      );
+      choroplethIncrementingId.current++;
```
can we store the layer id in a const and avoid storing the layer id in a ref? or do we need to increment it?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+          type: 'fill',
+          source: {
+            type: 'geojson',
+            data: propsRef.current.geojsonSource
```
what happens when these values change? whatever `useRef` returns is not updated right? Or does the behaviour of `useRef` change when you pass it a function?

dgieselaar replied:
Never mind this, `propsRef` will not change, only `.current` will so this code works 😅

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+  useEffect(() => {
+    const map = mapboxMap.current;
+    if (map && isMapReady.current) {
+      if (isEqual(mapboxStyleRef.current, mapboxStyle)) {
```
should this check be inversed?

dgieselaar commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,221 @@
+  const {
+    style,
+    getColorStyle,
+    mapboxStyle = 'https://tiles.maps.elastic.co/styles/osm-bright/style.json',
```
Given that we don't expect this to change during the lifespan of the component, maybe this should just be a hardcoded value? I think it would make the component simpler because you don't have to write any code that invalidates the map layer because of a style change (L184-206).

sqren commented:
Like @dgieselaar I think we should only make the properties we actually use configurable. I'm okay with hardcoding the rest (as constants) if that will simplify the implementation.

ogupte committed:
[APM] Adds chart for page load averages by country in RUM page-load view

ogupte committed:
[APM] Simplified and refined ChoroplethMap. Added legend labels.

ogupte committed:
- Replaced Map legend slices with smooth gradient
- fixed issue with map rendering multiple times
- renamed initial props to start with 'initial'
- added some more code comments

ogupte committed:
use correct i18n ids

sqren commented:
I'm seeing this odd behaviour when hovering over the map:
![geo-map-glitch](https://user-images.githubusercontent.com/209966/63890272-e7303980-c9e2-11e9-99f1-ca3df59c5ab4.gif)

sqren commented:
The tooltips can be a little hard to read when the text overlaps with continent names and such. I think adding a solid background color (white) would be good.

<img width="207" alt="Screen Shot 2019-08-28 at 22 28 18" src="https://user-images.githubusercontent.com/209966/63890407-31b1b600-c9e3-11e9-8f3d-cf0c72adf34b.png">
<img width="213" alt="Screen Shot 2019-08-28 at 22 28 05" src="https://user-images.githubusercontent.com/209966/63890408-31b1b600-c9e3-11e9-9117-6df078e21a72.png">

sqren commented:
**Legend**

I think the legend values should be rounded
<img width="1062" alt="Screen Shot 2019-08-28 at 22 30 36" src="https://user-images.githubusercontent.com/209966/63890708-d3390780-c9e3-11e9-9d21-ba1f479899be.png">

In this case the viz is based on a single doc. It's a bit of an edge case but perhaps the lower limit should always be 0ms? We do the same with the response time charts so might make sense for this too.
<img width="1109" alt="Screen Shot 2019-08-28 at 22 32 49" src="https://user-images.githubusercontent.com/209966/63890706-d3390780-c9e3-11e9-9d44-8ad1aa135b2c.png">

sqren commented:
**Legends**
I think a legend like this would be more clear

![](https://d33v4339jhl8k0.cloudfront.net/docs/assets/588089eddd8c8e484b24e90a/images/59f1fcc32c7d3a0bc56ab944/file-q420KeeWH8.png)

The current legend take up a lot of space at the bottom, but is not super clear. It also overlays some of the region names (South America and Oceania).

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ColorProgressionBar.tsx:
```diff
@@ -0,0 +1,33 @@
+
+import React from 'react';
+
+export const ColorProgressionBar: React.SFC<{
```
Maybe simply call this `Legend` since it's not really a reusable component but specifically designed for the purpose of being a legend.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/index.tsx:
```diff
@@ -150,51 +152,59 @@ export class TransactionCharts extends Component<TransactionChartProps> {
+            </EuiPanel>
+          </EuiFlexItem>
+        </EuiFlexGrid>
+        {transactionType === 'page-load' ? (
```
Perhaps `page-load` should be a constant?

sqren commented on x-pack/legacy/plugins/apm/server/lib/transactions/avg_duration_by_country/index.ts:
```diff
@@ -0,0 +1,77 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// import { flatten, sortByOrder } from 'lodash';
```
leftover?

sqren commented on x-pack/legacy/plugins/apm/server/lib/transactions/avg_duration_by_country/index.ts:
```diff
@@ -0,0 +1,77 @@
+      query: {
+        bool: {
+          filter: [
+            { term: { [SERVICE_NAME]: 'client' } },
```
Not everyone calls their application "client" ;)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,73 @@
+      <ColorProgressionBar
+        getColorStyle={getProgressionColor}
+        min={`${min}ms`}
+        max={`${max}ms`}
```
What if the durations are very high or very low? 

We recently [added support](https://github.com/elastic/kibana/pull/41640) in our formatters for minutes and hours after [requests from support and users](https://github.com/elastic/enhancements/issues/6640). 

All the formatters expect the time in microseconds, so instead of converting to millis you should just leave it in micros. And then use our formatters to format the value (eg. `getTimeFormatter`).

sqren replied:
yeah, I can't see your previous code because it's squashed but I seem to remember there was a transformation of the data somewhere. Instead I think you just need to use the formatters when outputting the durations. Like in the tooltip and on the legend.

sqren replied:
That being said: since it'll be ms in the large majority of cases this is probably not all that important.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+  return data.reduce(
+    ([min, max], { value }) => [Math.min(min, value), Math.max(max, value)],
+    [firstValue, firstValue]
+  );
```
What about making two generic functions that are a little easier to understand:

```js
const getMin = (items) => Math.min(...items.map(item => item.value))
const getMax = (items) => Math.max(...items.map(item => item.value))
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+  // converts input data to a scaled value between 0 and 1
+  const getValueScale = useCallback(
+    (value: number) => {
+      const [min, max] = getDataRange(data);
```
Since `getValueScale` is called within a loop `getDataRange` will be calculated for each element in the `data` array

sqren replied:
Yeah, I'd probably do something like:
```js
  const [min, max] = useMemo(() => getDataRange(data), [data]);
  const getValueScale = useCallback(
    (value: number) => (value - min) / (max - min),
    [min, max]
  );
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+
+  // initialization side effect, only runs once
+  useEffect(() => {
+    if (containerRef.current !== null) {
```
If you return early you can dedent the entire block below:
```js
if (containerRef.current === null) {
  return;
}
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+        updateHoverStateOnMouseoutRef.current
+      ) {
+        map.off('mousemove', updateHoverStateOnMousemoveRef.current);
+        map.off('mouseout', updateHoverStateOnMouseoutRef.current);
```
Shouldn't this happen during the cleanup phase? Our how do you remove the event handlers when the component unmounts?

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15062/)

ogupte committed:
- base color progression calc directly on euiColorPrimary
- check that a layer exists before querying features

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+        ...initialMapboxOptionsRef.current
+      });
+      mapboxMap.dragRotate.disable();
+      mapboxMap.touchZoomRotate.disableRotation();
```
Do you need to disable these after initialization? Can't you set them to `false` as part of the config above?

```js
const mapboxMap = new Map({
  dragRotate: false,
  touchZoomRotate: false
})

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,269 @@
+      if (map.getLayer(CHOROPLETH_LAYER_ID)) {
+        map.removeLayer(CHOROPLETH_LAYER_ID);
+      }
+      if (data.length) {
```
Again, return early. As a reader when I encounter this I have to prepare myself for to branches (I always scroll to the bottom to get an overview early). If you return early I immediately know there is only a single case left which makes it easier to reason about up front:

```js
if (data.length === 0) {
  return;
}
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+        popupRef.current &&
+        data.length &&
+        map.getLayer(CHOROPLETH_LAYER_ID)
+      ) {
```
Two comments; return early, and assign the condition to some descriptive variable. 

```js
const isSomethingSomething = map && popupRef.current && data.length && map.getLayer(CHOROPLETH_LAYER_ID)

if(!isSomethingSomething) {
 return
}

popupRef.current.setLngLat(event.lngLat);
...
```

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/15075/)

ogupte commented:
> **Legends**
> The current legend take up a lot of space at the bottom, but is not super clear. It also overlays some of the region names (South America and Oceania).

Discussed with @katrin-freihofner and @hbharding about this earlier, and the conclusion reached was to have the legend be a smooth gradient on the bottom, with labels at both ends.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+      <div ref={containerRef} style={{ height: 256, ...style }} />
+      <div style={{ display: 'none' }}>
+        <div ref={popupContainerRef}>
+          <ToolTip
```
Why is this renamed from `renderToolTip` to `ToolTip`? And why is `RegionMapChartToolTip` passed in as a prop in the first place? What about extract `RegionMapChartToolTip` to a separate file, and then use it directly?

sqren replied:
> I can definitely extract `RegionMapChartToolTip` to a separate file. I didn't at first because it's so small and specific. I like to keep component definitions of this nature very close to where they will be used. in the same source file even. But I'm not opposed to moving it.

Yeah, I think it makes sense to move it to it's own file and import it instead of passing it as a prop. Afaict we don't need to pass in different tooltips. If we eventually need it it'll be very easy to add.

sqren replied:
> Is a specific version of Chrome where you are seeing no background?

I'm on "Version 76.0.3809.100 (Official Build) (64-bit)"

sqren replied:
I'm still seeing this issue and @dgieselaar mentioned it in here: 
https://github.com/elastic/kibana/pull/43567#issuecomment-524542851

I think this is a blocker.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+  const [hoverState, setHoverState] = useState<{
+    geojsonProperties?: GeoJsonProperties;
+    data?: ChoroplethDataElement;
+  }>({});
```
Isn't this only used for the tooltip? If so, then what about calling it `tooltipState` and only set the things you need:
```js
  const [tooltipState, setTooltipState] = useState<{
  name: string
  value: number; 
  docCount: number;
  } | null>(null);
```

I think this will simplify a lot of other things.

sqren replied:
Right now we are the only consume so I think we should make things as simple as possible for ourselves.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/index.tsx:
```diff
@@ -0,0 +1,73 @@
+        initialGeojsonSource="https://vector.maps.elastic.co/files/world_countries_v1.geo.json?elastic_tile_service_tos=agree&my_app_name=ems-landing&my_app_version=7.2.0"
+        geojsonKeyProperty="iso2"
+        data={data}
+        renderTooltip={RegionMapChartToolTip}
```
Would it be possible to make all props except `data` constants inside `ChoroplethMap`?

```ts
<ChoroplethMap data={data}>
```

sqren replied:
I think we don't really know what "re-usable" means yet. Take a look at the embeddable Maps component. It's reusable but not to our use-case unfortunately. But it does have a lot of things we don't need ;)

So I think it's better to build something with the goal of making it as simple as possible, and then later _if_ we need it, it'll be easy to extend it and make it reusable.

I think `data` makes sense as a prop since it's not static. We still need to react to changes to `data` inside `useEffect` hooks and so forth. We don't need to do that for the others if they are constants inside the component.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+                    type: 'categorical',
+                    default: 'transparent'
+                  }
+                : 'transparent'
```
Can you extract it like:

```js
const stops = data.map(({ key, value }) => [
  key,
  getProgressionColor(getValueScale(value))
]);
const fillColor = data.length
  ? {
      property: geojsonKeyProperty,
      stops,
      type: 'categorical',
      default: 'transparent'
    }
  : 'transparent';

...

paint: {
  'fill-opacity': 0.75,
  'fill-color': fillColor
}
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+                geojsonProperties &&
+                key === geojsonProperties[geojsonKeyProperty]
+            );
+            setHoverState({ geojsonProperties, data: matchedData });
```
You use `data` both for the array of items and for the single item. Can you find a noun that makes it clear whether it's plural or singular? `items` and `item` would be fine.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/RegionMapChart/ChoroplethMap.tsx:
```diff
@@ -0,0 +1,276 @@
+            setHoverState({ geojsonProperties, data: matchedData });
+          }
+        } else {
+          setHoverState({});
```
I think resetting to an object instead of `null` points to some other issues with the data structure. I think this would be better:

```js
setHoverState(null);
```

sqren commented on x-pack/legacy/plugins/apm/server/lib/transactions/avg_duration_by_country/index.ts:
```diff
@@ -0,0 +1,77 @@
+  };
+
+  const resp = await client.search(params);
+  if (resp.aggregations) {
```
```js
if (!resp.aggregations) {
  return [];
}
```

ogupte committed:
Addressed code review feedback

ogupte committed:
- fixes issue where min/max was not a finite value
- cleans up mouseover handler, which updates on state changes
- formats doc count for display
- style improvements

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/33/)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -0,0 +1,278 @@
+
+    // only scroll zoom when key is pressed
+    const canvasElement = mapboxMap.getCanvas();
+    canvasElement.addEventListener('wheel', controlScrollZoomOnWheel);
```
Instead of attaching the listener to the canvas element it looks like you can add it to the `mapboxMap` like you do with mouse events. That way we don't have to worry about cleanup either (if I understood you correctly that this was handled internally by mapbox)

```js
    // only scroll zoom when key is pressed
    mapboxMap.on('wheel', event => {
      if (event.originalEvent.ctrlKey) {
        return;
      }

      if (event.originalEvent.metaKey) {
        return;
      }

      if (event.originalEvent.altKey) {
        return;
      }

      event.preventDefault();
    });
```

sqren replied:
That's odd. I commented out:
```
canvasElement.addEventListener('wheel', controlScrollZoomOnWheel);
```

Which turned on the scroll zoom. 
And was then able to disable it again with:

```js
    mapboxMap.on('wheel', event => {
      if (
        event.originalEvent.ctrlKey ||
        event.originalEvent.metaKey ||
        event.originalEvent.altKey
      ) {
        return;
      }

      event.preventDefault();
    });
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -0,0 +1,278 @@
+    mapboxMap.on('mousemove', (...args) =>
+      updateTooltipStateOnMousemoveRef.current(...args)
+    );
+    mapboxMap.on('mouseout', updateTooltipStateOnMouseout);
```
Instead of defining `updateTooltipStateOnMouseout` outside the `useEffect` can't you do this? This way we don't have to pass it as a dependency and argument types are inferred automatically

```js
    mapboxMap.on('mouseout', event => {
      setTooltipState(null);
    });
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/PageLoadCharts/index.tsx:
```diff
@@ -0,0 +1,36 @@
+              )}
+            </span>
+          </EuiTitle>
+          <ChoroplethMap data={data} />
```
Do you mind making it:

```js
<ChoroplethMap items={data} />
```

so we always have a plural value inside `ChoroplethMap`.

Btw. We haven't done this much but I think that specialized data fetcher hook (like `useAvgDurationByCountry`) should return a better named variable than `data` for their data. Again, `items` in this case is fine.

`useFetcher` will always return `data` through since that is completely generic.

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -0,0 +1,278 @@
+  // and replaces the old one stored in `updateTooltipStateOnMousemoveRef`
+  useEffect(() => {
+    const updateTooltipStateOnMousemove = (
+      event: mapboxgl.MapMouseEvent & mapboxgl.EventData
```
I could be wrong but it looks like you only need `MapMouseEvent` type here:
```js
event: mapboxgl.MapMouseEvent
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -0,0 +1,278 @@
+
+  const updateTooltipStateOnMouseout = useCallback(
+    (event: mapboxgl.MapMouseEvent & mapboxgl.EventData) => {
+      enableScrollZoom.current = false;
```
What is `enableScrollZoom` for? I don't see it used anywhere

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/ChoroplethToolTip.tsx:
```diff
@@ -0,0 +1,44 @@
+export const ChoroplethToolTip: React.SFC<{
+  name?: string;
+  value?: number;
+  docCount?: number;
```
Can we make all of these required so we don't have to make the `null` checks inside the tooltip?
```js
  name: string;
  value: number;
  docCount: number;
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/Legend.tsx:
```diff
@@ -0,0 +1,41 @@
+      }}
+    >
+      <LegendLabel>{asTime(min)}</LegendLabel>
+      <LegendLabel>{asTime(max)}</LegendLabel>
```
When using `asTime` you use a new formatter every time. So if `min` could be formatted as milliseconds while `max` is formatted as seconds. Often you want to use the same formatter throughout a component so units don't change. In that case I'd probably base the formatter on the `max` value.

You can do this:

```ts
const formatter = getTimeFormatter(max)

// ---

<LegendLabel>{formatter(min)}</LegendLabel>
<LegendLabel>{formatter(max)}</LegendLabel>
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -0,0 +1,278 @@
+    const maxValue = getMax(data);
+    return [
+      Number.isFinite(minValue) ? minValue : 0,
+      Number.isFinite(maxValue) ? maxValue : 0
```
Isn't this only relevant when `data`/`items` is empty? And in those cases do we even want to show the legend?

sqren replied:
Ah, good point. Didn't realized that. I think that's okay for now - the legend currently looks like this without any data so I think it's better to not show it:

<img width="1106" alt="Screen Shot 2019-08-29 at 09 18 08" src="https://user-images.githubusercontent.com/209966/63918939-04debc80-ca3e-11e9-9c84-f48b9b3026b4.png">

ogupte committed:
addressed PR feedback & updated renovate.json5

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/47/)

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -263,16 +247,21 @@ export const ChoroplethMap: React.SFC<Props> = props => {
-      <Legend getColorStyle={getProgressionColor} min={min} max={max} />
+      {hasMinMax ? (
+        <Legend getColorStyle={getProgressionColor} min={min} max={max} />
+      ) : null}
```
I think what we really want is to say: hide the legend if there are no items. `min` and `max` should be defined if there are items (if not we should clean up the list before passing it to `ChoroplethMap` imo)

```js
      {items.length > 0 && (
        <Legend getColorStyle={getProgressionColor} min={min} max={max} />
      )}
```

sqren commented on x-pack/legacy/plugins/apm/public/components/shared/charts/TransactionCharts/ChoroplethMap/index.tsx:
```diff
@@ -263,16 +247,21 @@ export const ChoroplethMap: React.SFC<Props> = props => {
       <div style={{ display: 'none' }}>
         <div ref={popupContainerRef}>
-          <ChoroplethToolTip {...tooltipState} />
+          <ChoroplethToolTip {...toolTipProps} />
```
Instead of setting default values, what about not showing the tooltip if there is no tooltip data to show?
```js
{tooltipState && <ChoroplethToolTip {... tooltipState} />}
```

sorenlouv reviewed (changes requested):
@ogupte From a code review perspective this is in good shape - I just added a few more comments but nothing blocking. Good job 👍 

From an end-user perspective I think there is blocker. There seems to be an issue with the tooltips. Both @dgieselaar and I are seeing them without the solid white background, and for me they "stutter" when hovering over certain countries (in this case the US - don't take it personal 🇺🇸😀).

![](https://user-images.githubusercontent.com/209966/63890272-e7303980-c9e2-11e9-99f1-ca3df59c5ab4.gif)

spalger reviewed (approved):
Renovate changes LGTM

dgieselaar commented:
@ogupte	I'll excuse myself from further reviewing 😀 don't want to add a lot of noise given the time constraints, and in general I agree with @sqren's feedback.

ogupte committed:
- Removed the Legend from the ChoroplethMap
- Only render tooltip when there's data

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/169/)

ogupte committed:
- fix hover state not clearing properly
- add better typing around Geojson propertier for world countries

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/228/)

ogupte committed:
added missing css import

sorenlouv reviewed (approved):
🌎

nik9000 commented:
@elasticmachine run elasticsearch-ci/docs

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pull-request/263/)
